File: API/Services/UserService.cs
using API.Models;
using API.Repositories;

namespace API.Services
{
    public class UserService
    {
        private readonly UserRepository _repository;
        private readonly TokenService _tokenService;

        public UserService(TokenService tokenService)
        {
            _repository = new UserRepository();
            _tokenService = tokenService;
        }

        public dynamic ValidateUser(UserLogin login)
        {
            var user = _repository.ValidateUser(login.user, login.password);
            if (user == null) throw new Exception("Invalid credentials");

            return new UserLoginResponse
            {
                role = user.role,
                name = user.name,
                token = _tokenService.CreateToken(user.name)
            };
        }

        public List<dynamic> GetAll()
        {
            return _repository.SelectAll();
        }

        public dynamic GetById(long id)
        {
            if (id <= 0) throw new Exception("Id can't be null");
            return _repository.SelectById(id);
        }

        public dynamic Insert(UserInsert obj)
        {
            if (string.IsNullOrEmpty(obj.email)) throw new Exception("Email can't be null");
            if (string.IsNullOrEmpty(obj.name)) throw new Exception("Name can't be null");
            if (string.IsNullOrEmpty(obj.password)) throw new Exception("Password can't be null");
            return _repository.Insert(obj);
        }

        public dynamic Update(UserUpdate obj)
        {
            if (obj.id <= 0) throw new Exception("Id can't be null");
            if (string.IsNullOrEmpty(obj.email)) throw new Exception("Email can't be null");
            if (string.IsNullOrEmpty(obj.name)) throw new Exception("Name can't be null");
            
            dynamic exists = _repository.SelectById(obj.id);
            int rows = 0;
            if (exists.id != -1) rows = _repository.UpdateById(obj);
            return rows;
        }

        public dynamic Delete(long id)
        {
            if (id <= 0) throw new Exception("Id can't be null");
            dynamic exists = _repository.SelectById(id);
            int rows = 0;
            if (exists.id != -1) rows = _repository.DeleteById(id);
            return rows;
        }
    }
}

File: Frontend/src/app/core/services/UserService.ts
import { BehaviorSubject, map } from 'rxjs';
import { Injectable } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { Constants } from './Constants';
import { User, UserLogin, UserLoginResponse, UserInsert, UserUpdate } from '../models/User';

@Injectable({
  providedIn: 'root',
})
export class UserService {
  private BASE_URL: string = Constants.USER;
  private currentUserSource = new BehaviorSubject<any | null>(null);
  currentUser$ = this.currentUserSource.asObservable();

  getCurrentUser = () => sessionStorage.getItem('name');
  getCurrentAccessLevel = () => sessionStorage.getItem('roles');
  getCurrentToken = () => sessionStorage.getItem('token');
  getCurrentUserId = () => sessionStorage.getItem('userId');

  constructor(private http: HttpClient) { }

  Login = (user: UserLogin) => this.http.post<UserLoginResponse>(`${this.BASE_URL}/Validate`, user).pipe(
    map((user: UserLoginResponse) => {
      sessionStorage.setItem('roles', user.role);
      sessionStorage.setItem('name', user.name);
      sessionStorage.setItem('token', user.token);
    }));

  GetList = () => this.http.get<User[]>(`${this.BASE_URL}/Read`);
  
  GetById = (id: number) => this.http.get<User>(`${this.BASE_URL}/Read/${id}`);
  
  Insert = (obj: UserInsert) => this.http.post(`${this.BASE_URL}/Create`, obj);
  
  UpdateById = (obj: UserUpdate) => this.http.put(`${this.BASE_URL}/UpdateById`, obj);
  
  DeleteById = (id: number) => this.http.delete(`${this.BASE_URL}/DeleteById/${id}`);

  Logout = () => {
    sessionStorage.clear();
  }
}